File: src/main/java/com/github/hirsivaja/ip/IpPacket.java
package com.github.hirsivaja.ip;

import com.github.hirsivaja.ip.ethernet.EthernetPayload;
import com.github.hirsivaja.ip.ipv4.Ipv4Header;
import com.github.hirsivaja.ip.ipv4.Ipv4Packet;
import com.github.hirsivaja.ip.ipv6.Ipv6Header;
import com.github.hirsivaja.ip.ipv6.Ipv6Packet;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.Logger;

public sealed interface IpPacket extends EthernetPayload permits Ipv4Packet, Ipv6Packet {
    Logger logger = Logger.getLogger("IpPacket");
    IpHeader header();

    @Override
    default byte[] toBytes() {
        ByteBuffer out = ByteBuffer.allocate(length());
        encode(out);
        byte[] outBytes = Arrays.copyOfRange(out.array(), 0, out.rewind().remaining());
        if(logger.isLoggable(Level.FINE)) {
            logger.log(Level.FINE, "IP packet as byte array:\n\t{0}", IpUtils.printHexBinary(outBytes));
        }
        return outBytes;
    }

    default String toByteString() {
        return IpUtils.printHexBinary(toBytes());
    }

    static IpPacket fromBytes(byte[] ipPayload) {
        if(logger.isLoggable(Level.FINE)) {
            logger.log(Level.FINE, "Creating an IP packet from:\n\t{0}", IpUtils.printHexBinary(ipPayload));
        }
        return decode(ByteBuffer.wrap(ipPayload));
    }

    static IpPacket fromByteString(String ipPayload) {
        return fromBytes(IpUtils.parseHexBinary(ipPayload));
    }

    static IpPacket decode(ByteBuffer in) {
        return decode(in, true);
    }

    static IpPacket decode(ByteBuffer in, boolean ensureChecksum) {
        in.mark();
        byte version = (byte) (in.get() >>> Ipv4Header.VERSION_SHIFT);
        in.reset();
        return switch (version) {
            case Ipv4Header.VERSION -> Ipv4Packet.decode(in, ensureChecksum);
            case Ipv6Header.VERSION -> Ipv6Packet.decode(in, ensureChecksum);
            default -> throw new IllegalArgumentException("Not an IP data");
        };
    }
}

File: src/main/scala/com/github/hirsivaja/ip/IpPacket.scala
package com.github.hirsivaja.ip

import com.github.hirsivaja.ip.header.{IpHeader, Ipv4Header, Ipv6Header}
import com.github.hirsivaja.ip.payload.{IpPayload, Ipv4Payload, Ipv6Payload}

import java.nio.ByteBuffer
import java.util.HexFormat

sealed trait IpPacket {
  val header: IpHeader
  val payload: IpPayload
  val length: Int = header.length + payload.length

  def encode(out: ByteBuffer): Unit = {
    header.encode(out)
    payload.encode(out)
  }

  def toBytes: Array[Byte] = {
    val out = ByteBuffer.allocate(length)
    encode(out)
    out.array().array.clone()
  }

  def toByteString: String = {
    HexFormat.of.formatHex(toBytes).toUpperCase
  }
}

object IpPacket {
  def fromBytes(ipPacket: Array[Byte]): IpPacket = {
    decode(ByteBuffer.wrap(ipPacket))
  }

  def fromByteString(ipPacket: String): IpPacket = {
    fromBytes(HexFormat.of.parseHex(ipPacket))
  }

  def decode(in: ByteBuffer): IpPacket = {
    decode(in, true)
  }

  def decode(in: ByteBuffer, ensureChecksum: Boolean): IpPacket = {
    in.mark
    val version = (in.get >>> Ipv4Header.VersionShift).toByte
    in.reset
    version match {
      case Ipv4Header.Version => Ipv4Packet.decode(in, ensureChecksum)
      case Ipv6Header.Version => Ipv6Packet.decode(in, ensureChecksum)
      case _ => throw new IllegalArgumentException("Not an IP data")
    }
  }
}

final case class Ipv4Packet(override val header: Ipv4Header, override val payload: Ipv4Payload) extends IpPacket {}

object Ipv4Packet {
  def decode(in: ByteBuffer, ensureChecksum: Boolean): Ipv4Packet = {
    val header = Ipv4Header.decode(in, ensureChecksum)
    val payload = Ipv4Payload.decode(in, ensureChecksum, header)
    new Ipv4Packet(header, payload)
  }
}

final case class Ipv6Packet(override val header: Ipv6Header, override val payload: Ipv6Payload) extends IpPacket {}

object Ipv6Packet {
  def decode(in: ByteBuffer, ensureChecksum: Boolean): Ipv6Packet = {
    val header = Ipv6Header.decode(in)
    val payload = Ipv6Payload.decode(in, ensureChecksum, header)
    new Ipv6Packet(header, payload)
  }
}
